### ISSUE TITLE
api.tools.parse_requests mutating the xhrParams

### ISSUE THREAD

**Issue by kinkoazc** [KartikTalwar/gmail.js#662]
Hello,

I came across an issue involving the use of `gmail.js` library by 2(or more) currently installed extensions.
I think that they come into conflict around the mutable changes done to the `params` variable, in the `api.tools.parse_requests` method.

Scenario:
1. chrome extension no. 1 patches the `send` method, `api.tools.patch(win.XMLHttpRequest.prototype.send`;
2. chrome extension no. 2 also patches the `send` method(now, already patched by chrome extension no. 1), `api.tools.patch(win.XMLHttpRequest.prototype.send`;
3. a XHR request gets made, with `this.xhrParams` initially as `{method: 'POST', url: 'https://mail.google.com/sync/u/0/i/s?hl=en&c=21'}`;
4. chrome extension no. 2's patch comes into play, executes `api.tools.parse_requests`; now `this.xhrParams` looks like `{method: 'POST', url_raw: 'https://mail.google.com/sync/u/0/i/s?hl=en&c=21', body_params: {...}, url: {hl: 'en', c: '21'}, body: '...'}`;
5. chrome extension no. 2's patch sends the request, `var out = orig.apply(this, arguments);`;
6. now chrome extension no. 1 patch comes into play, receiving, as `this.xhrParams`, the above `{method: 'POST', url_raw: 'https://mail.google.com/sync/u/0/i/s?hl=en&c=21', body_params: {...}, url: {hl: 'en', c: '21'}, body: '...'}`;
7. in chrome extension no. 1's `api.tools.parse_requests` method `this.xhrParams` will come out as `{method: 'POST', url_raw: {hl: 'en', c: '21'}, body_params: {...}, url: {}, body: '...'}`, which will mess with the subsequent library logic, e.g. what happens in `api.tools.parse_request_payload`.

Is this behavior correct/known? Thank you.

**Comment by josteink** [comment]
Oh bugger. 

I really did my best to rewrite Gmail.js to be “stackable” without causing issues with concurrent extensions using it. 

Guess there’s more work which needs to be done then.

Not sure when I will have time to go in depth into this. If you can find a way to avoid this causing issues, I’m open to advice. 

And obviously: PRs are always welcome 🙂

**Comment by josteink** [comment]
Hey there and thanks for the detailed report. 

This issue was not really known, to be honest. 

In other words: Oh bugger. 

I really did my best to rewrite Gmail.js to be “stackable” without causing issues with concurrent extensions using it. 

I guess this mostly affects extensions which modifies the requests in event handlers? Or does it affect all? 

Either way: Guess there’s more work which needs to be done then.

Not sure when I will have time to go in depth into this. If you can find a way to avoid this causing issues, I’m open to advice. 

And obviously: PRs are always welcome 🙂

**Comment by kinkoazc** [comment]
Hi. Thank you for your great work and contribution to this library. Also thank you for taking a look at and confirming the issue.
Sure, I will create a PR this week.

**Comment by josteink** [comment]
If we use a function like time or random, perhaps we can assign a instance-ID per GmailJS instance, and use that when assigning variables like this?

    this.xhr_params_gmailjs[this.instance_id] = xhr_params;

Then different instances will only “see” variables belonging to them?

Shouldn’t that work?

**Comment by kinkoazc** [comment]
Yes. What I did to address the issue, while waiting for a library solution, is to do a second patching of the `open` and `parse_request` methods:
``` js
        this.email = new GmailJs.Gmail($);

        const win = this.email.helper.get_xhr_window();
        const xhrParamsKey = Symbol('some key, for debugging purposes'));
        if (win) {
            this.email.tools.patch(win.XMLHttpRequest.prototype.open, (orig) => {
                win.XMLHttpRequest.prototype.open = function (method, url, async, user, password) {
                    let out = orig.apply(this, arguments);
                    Object.defineProperty(this, xhrParamsKey, {
                        value: Object.freeze({
                            method: method.toString(),
                            url: url.toString()
                        })
                    });
                    return out;
                };
            });
        }
        const originalParseRequestsMethod = this.email.tools.parse_requests;
        this.email.tools.parse_requests = (params, xhr) => {
            if (params && typeof params.url !== 'string') {
                if (
                    xhr
                    && xhr[xhrParamsKey]
                    && xhr[xhrParamsKey].url
                ) {
                    params.url = xhr[xhrParamsKey].url;
                    delete params.url_raw;
                    delete params.body_params;
                } else if (
                    params.url_raw
                    && typeof params.url_raw === 'string'
                ) {
                    params.url = params.url_raw;
                    delete params.url_raw;
                    delete params.body_params;
                }
            }

            return originalParseRequestsMethod(params, xhr);
        };
```
 Something on these lines should fix this issue I think(while still keeping the `this.xhrParams` creation and use, for backwards compatibility), since the runtime order is `open, extension 2 -> open, extension 1 -> send, extension 2 -> send, extension 1`. I'll have to test the impact on the rest of the library's API, as well as on other/old versions of the library, being part of the same runtime stack. I'm also thinking of trying to clean the (future legacy) `this.xhrParams` object for the next(wrapped/inner) instance of the library.

**Comment by josteink** [comment]
Closed via #666 (nice PR number :imp: )!

Should now we be published on npmjs in `v1.0.16`.

**Comment by kinkoazc** [comment]
Yes, was hiding in the details :-)
Thank you as well for your help on addressing this issue!
